perf/src/sigverify.rs

```diff
@@ -113,6 +122,66 @@ fn verify_packet(packet: &mut Packet) {
         return;
     }
 
+    if reject_non_vote {
+        if packet_offsets.sig_len != 1 {
+            packet.meta.discard = true;
+            return;
+        }
+
+        let mut reject_pubkey_start = pubkey_start;
+        let mut vote_index = None;
+        for i in 0..packet_offsets.pubkey_len {
+            let pubkey_end = reject_pubkey_start.saturating_add(size_of::<Pubkey>());
+            if &packet.data[pubkey_start..pubkey_end] == solana_vote_program::id().as_ref() {
+                vote_index = Some(i);
+                break;
+            }
+            reject_pubkey_start = pubkey_end;
+        }
+
+        if vote_index.is_none() {
+            packet.meta.discard = true;
+            return;
+        }
+
+        let pubkeys_end = pubkey_start
+            .saturating_add(size_of::<Pubkey>().saturating_mul(packet_offsets.pubkey_len as usize));
+        let hash_end = pubkeys_end.saturating_add(size_of::<Hash>());
+        let num_instructions_offset = hash_end;
+
+        if hash_end.saturating_add(1) >= packet.meta.size {
+            packet.meta.discard = true;
+            return;
+        }
+
+        let (instructions_len, instructions_size) =
+            match decode_shortu16_len(&packet.data[num_instructions_offset..]) {
+                Ok((len, size)) => (len, size),
+                Err(_) => {
+                    packet.meta.discard = true;
+                    return;
+                }
+            };
+
+        let instruction0_start = hash_end.saturating_add(instructions_size);
+        if instruction0_start > packet.meta.size {
+            packet.meta.discard = true;
+            return;
+        }
+
+        let vote_index = vote_index.unwrap();
+        // First instruction should be the vote key
+        if packet.data[instruction0_start] as u32 != vote_index {
+            packet.meta.discard = true;
+            return;
+        }
+
+        if instructions_len != 1 {
```

Comment:
Should be faster to first check that there's only one instruction as you're doing here and _then_ look up that ix's program id index than it is to scan through each pubkey checking it against the vote program id.